interfaces/connection_test.go

```diff
@@ -0,0 +1,213 @@
+
+	attrs, err := attrData.StaticAttrs()
+	c.Assert(err, IsNil)
+	c.Assert(attrs, DeepEquals, map[string]interface{}{
+		"attr": "value",
+	})
+}
+
+func (s *connSuite) TestDynamicSlotAttrs(c *C) {
+	attrs := map[string]interface{}{
+		"foo": "bar",
+	}
+	attrData, _ := NewConnectedSlot(s.slot, attrs)
+	c.Assert(attrData, NotNil)
+
+	val, err := attrData.Attr("foo")
+	c.Assert(err, IsNil)
+	c.Assert(val, Equals, "bar")
+
+	val, err = attrData.Attr("attr")
+	c.Assert(err, IsNil)
+	c.Assert(val, Equals, "value")
+
+	_, err = attrData.Attr("unknown")
+	c.Assert(err, NotNil)
+	c.Assert(err, ErrorMatches, `attribute "unknown" not found`)
+
+	attrs, err = attrData.Attrs()
+	c.Assert(err, IsNil)
+	c.Assert(attrs, DeepEquals, map[string]interface{}{
+		"foo": "bar",
+	})
+}
+
+func (s *connSuite) TestDynamicPlugAttrs(c *C) {
+	attrs := map[string]interface{}{
+		"foo": "bar",
+	}
+	plug, _ := NewConnectedPlug(s.plug, attrs)
+	c.Assert(plug, NotNil)
+
+	val, err := plug.Attr("foo")
+	c.Assert(err, IsNil)
+	c.Assert(val, Equals, "bar")
+
+	val, err = plug.Attr("attr")
+	c.Assert(err, IsNil)
+	c.Assert(val, Equals, "value")
+
+	_, err = plug.Attr("unknown")
+	c.Assert(err, NotNil)
+	c.Assert(err, ErrorMatches, `attribute "unknown" not found`)
+
+	attrs, err = plug.Attrs()
+	c.Assert(err, IsNil)
+	c.Assert(attrs, DeepEquals, map[string]interface{}{
+		"foo": "bar",
+	})
+}
+
+func (s *connSuite) TestDynamicPlugAttrsNotInitialized(c *C) {
+	slot, _ := NewConnectedPlug(s.plug, nil)
+	c.Assert(slot, NotNil)
+
+	_, err := slot.Attr("foo")
+	c.Assert(err, NotNil)
+	c.Assert(err, ErrorMatches, `attribute "foo" not found`)
+
+	_, err = slot.Attrs()
+	c.Assert(err, ErrorMatches, `dynamic attributes not initialized`)
+}
+
+func (s *connSuite) TestOverwriteStaticAttrError(c *C) {
+	attrs := map[string]interface{}{}
+
+	plug, _ := NewConnectedPlug(s.plug, attrs)
+	c.Assert(plug, NotNil)
+	slot, _ := NewConnectedSlot(s.slot, attrs)
+	c.Assert(slot, NotNil)
+
+	err := plug.SetAttr("attr", "overwrite")
+	c.Assert(err, NotNil)
+	c.Assert(err, ErrorMatches, `attribute "attr" cannot be overwritten`)
+
+	err = slot.SetAttr("attr", "overwrite")
+	c.Assert(err, NotNil)
+	c.Assert(err, ErrorMatches, `attribute "attr" cannot be overwritten`)
+}
+
+func (s *connSuite) TestDynamicSlotAttrsNotInitialized(c *C) {
+	slot, _ := NewConnectedSlot(s.slot, nil)
+	c.Assert(slot, NotNil)
+
+	_, err := slot.Attr("foo")
+	c.Assert(err, NotNil)
+	c.Assert(err, ErrorMatches, `attribute "foo" not found`)
+
+	_, err = slot.Attrs()
+	c.Assert(err, ErrorMatches, `dynamic attributes not initialized`)
+}
+
+func (s *connSuite) TestCopyAttributes(c *C) {
+	orig := map[string]interface{}{
+		"a": "A",
+		"b": true,
+		"c": int(100),
+		"d": []interface{}{"x", "y", true},
+		"e": map[string]interface{}{
+			"e1": "E1",
+		},
+	}
+
+	cpy, err := CopyAttributes(orig)
+	c.Assert(err, IsNil)
+	// verify that int is converted into int64
+	c.Check(reflect.TypeOf(cpy["c"]).Kind(), Equals, reflect.Int64)
```

Comment:
Similarly, that feels strange. It feels like a very awkward thing to be doing on the copy operation. The value shouldn't reach that internal dictionary with a wrong type, and it should break at set time rather than penalizing those reading the broken value.

Reply:
Break or be converted. In this particular case (int => int64) automatically converting on set sounds better.